selinux: Show local modifications

marusak opened the pull request:
![zebra](https://user-images.githubusercontent.com/12330670/57528401-8d700a80-7332-11e9-88aa-ce234fdec3a4.png)
![sranda_vyskakovacia](https://user-images.githubusercontent.com/12330670/57519592-f13b0900-731b-11e9-8ea1-d9264aa7bfe5.png)

Mostly Fixes #8657

There is no ansible script, this is gonna be follow up for sure.
This PR for now only shows booleans in normal readable form, others are just dumped from `semange export`. Each type needs to have its own parsing function and then just add this function into the map of functions. Some of these types could be introduced in this PR, but I also expect most of them as follow up.

@andreasn could you take a look and let me know if design is right or you would change something?
Thank you!

andreasn commented on pkg/lib/cockpit-components-modifications.css:
```diff
@@ -0,0 +1,30 @@
+}
+
+.modifications-caption a.modifications-export {
+    font-size: 11px;
```
This feels very small to me. Probably want to make this 12px at minimum.

andreasn commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,152 @@
+                </Modal.Body>
+                <Modal.Footer>
+                    <Button bsStyle='default' className='btn-cancel' onClick={this.props.onClose}>
+                        { _("Cancel") }
```
We're not really cancelling any action here, so _Close_ is probably a better verb.

andreasn commented on pkg/lib/cockpit-components-modifications.css:
```diff
@@ -0,0 +1,30 @@
+.modifications-table {
+    border: solid 2px #f1f1f1;
```
I keep having issues with fuzzy exports in Inkscape, so it might have looked like 2px on the mockup.
I intended the style to be like dialog-list-ct, like in the storage dialogs
![Screenshot from 2019-05-10 18-42-34](https://user-images.githubusercontent.com/209895/57542981-6544d380-7353-11e9-98da-35d21f8ebe75.png)

garrett replied:
Here's the reason for fuzzy exports:

Inkscape switched from 90 DPI (which is weird, but it used it for _years_) to 96 DPI. Make sure you have the correct DPI in the export dialog.

Inkscape warns you when you open an older file and lets you fix it... but exporting settings might not be obvious.

If in doubt, you can always put int the pixel size you expect and it'll change the DPI accordingly.

marusak commented:
Thank you @andreasn for the review! I fixed your comments.
While on it I noticed that when I login without `Reuse my password for privileged tasks` I cannot see any modifications - this makes sense since running `semanage export` as normal user is not allowed. It will show `No Local Modifications` which is rather misleading. How do you suggest showing that user does not have privileges? We need to distinguish between having privileges but no modification are present and not having privileges.

andreasn commented:
_The user **username** is not permitted to..._ is the wording we use for the tooltips, so something like that should work.
![Screenshot from 2019-05-13 10-47-27](https://user-images.githubusercontent.com/209895/57608104-ab28b400-756c-11e9-87d3-fcf4af7713cb.png)

What about _The user **username** is not permitted to view local modifications_ ?

marusak commented:
Like this @andreasn ?

When user is not permitted:
![nahnah](https://user-images.githubusercontent.com/12330670/57612078-6f91e800-7574-11e9-8cd0-d3f77d1eb67a.png)


When user is permitted but nothing to show:
![permitedbutno](https://user-images.githubusercontent.com/12330670/57611887-1c1f9a00-7574-11e9-927b-b6e13cffdd63.png)

When user is permitted and there is something to show:
![localmodificationpresent](https://user-images.githubusercontent.com/12330670/57611914-280b5c00-7574-11e9-83da-606ff2c4166e.png)

marusak commented on pkg/selinux/selinux-client.js:
```diff
@@ -42,6 +45,9 @@ var status = {
  *   - configEnforcing: boolean (mode the system is configured to boot in, may differ from current mode)
+ *   - shell:           Output of `semanage export`
+ *   - modifications:   List of all local modifications in selinux policy
+ *   - permitted:       Set to force is user is not allowed to see local modifications
```
Set to false... - fix on next push

martinpitt replied:
... and s/is/if/

martinpitt commented on pkg/lib/cockpit-components-modifications.css:
```diff
@@ -0,0 +1,37 @@
+    margin-bottom: 25px;
+    min-width: 65%;
+    width: 100%;
+}
```
As this is quite a lot of custom CSS, I propose to sort this PR after the PF4 rework (PR #11987), and then update the CSS accordingly. Perhaps we'll then also need less of it :-)

marusak replied:
We maybe should introduce some label and give it to all open PRs that touch css (or possibly even html) to know, that those should be checked if it complies with PF4.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+        super(props);
+
+        this.state = {
+            key: 1
```
At first sight this sounds like the currently active tab number? There is no `state.key` anywhere, so supposedly `ModificationsExportDialog` handles that internally? This should get a comment what this is.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+                    <TabContainer id="basic-tabs-pf" defaultActiveKey={1}>
+                        <React.Fragment>
+                            <Nav bsClass="nav nav-tabs nav-tabs-pf" onSelect={this.handleSelect}>
+                                <NavItem eventKey={1}>
```
OOI, do these have to be integers, or could they be strings like `shell` and `ansible`?

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+                                </NavItem>
+                                {this.props.ansible &&
+                                    <NavItem eventKey={2}>
+                                        {_("Ansible script")}
```
Let's call it Ansible Playbook".

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+                        <React.Fragment>
+                            <Nav bsClass="nav nav-tabs nav-tabs-pf" onSelect={this.handleSelect}>
+                                <NavItem eventKey={1}>
+                                    {_("Shell script")}
```
I think we usually use Title Case for these.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+    onClose: PropTypes.func.isRequired,
+};
+
+/* Display list of modification in human readable format
```
modification"s"

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+/* Display list of modification in human readable format
+ *
+ * Also show `View automation script` button which opens dialog in which different
+ * scripts are available. With these script it is possible to get other macchines
```
script"s" ... possible to apply the same configuration to other machines

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+    render() {
+        var emptyRow = null;
+        if (!this.props.entries || this.props.entries.length === 0) {
+            var markup = {
```
`const` here, please. `var` is broken unless in `function` scope.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+                __html: this.props.emptyCaption,
+            };
+            emptyRow = <thead className={"listing-ct-empty"}>
+                <tr className="modification-row"><td dangerouslySetInnerHTML={ markup } /></tr>
```
Let's please avoid the dangerouslySetInnerHTML. Rather, let callers pass a component, or just use a generic string like "No modifications.". We can make this more flexible if/when needed later.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+
+        return (
+            <React.Fragment>
+                <ModificationsExportDialog show={this.state.showDialog} shell={ this.props.shell } ansible={ this.props.ansible } onClose={ () => this.setState({ showDialog: false }) } />
```
Please be consistent about the spacing in `{}`. I usually avoid the space for single values/constants, and have spaces for more complex expressions for easier readability.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+                <table className={ "listing-ct modifications-table" }>
+                    <caption className="cockpit-caption">
+                        <div className="modifications-caption">
+                            { this.props.title }
```
Why isn't this using a standard `<h3>`? Probably more a11y and we could lose the custom CSS?

marusak replied:
Because it is in `<caption>` element, size of the title is set up through `listing.less` style. Purpose of this is to be able to show the "View automation script" in the same line and to make it smaller than the caption itself.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,158 @@
+    title: PropTypes.string.isRequired,
+    emptyCaption: PropTypes.string.isRequired,
+    entries: PropTypes.arrayOf(PropTypes.string).isRequired,
+    shell: PropTypes.string,
```
shell is required by ModificationsExportDialog.

martinpitt commented on pkg/selinux/selinux-client.js:
```diff
@@ -96,6 +102,66 @@ export function init(statusChangedCallback) {
         );
     };
 
+    var parseBoolean = function(result, item) {
```
This is weird syntax. Why not `function parseBoolean()`?

martinpitt commented on pkg/selinux/selinux-client.js:
```diff
@@ -96,6 +102,66 @@ export function init(statusChangedCallback) {
+        // Use `echo '~~~~~'` as separator, so we don't need to execute multiple commands
+        let script = "semanage export";
+        manageditems.forEach(item => { script += " && echo '~~~~~' && semanage " + item + " --list -C --noheading" });
+        cockpit.spawn(["/bin/sh", "-c", script], { err: 'message', environ: [ "LC_ALL=C" ], superuser: "try" }).then(
```
You can use `cockpit.script` for this. Also, please set `LC_MESSAGES` only, so that this doesn't damage non-ASCII file names.

Put the `.then()` on a new line and use an arrow instead of `function(output)` so that this becomes bit tidier.

martinpitt commented on pkg/selinux/selinux-client.js:
```diff
@@ -96,6 +102,66 @@ export function init(statusChangedCallback) {
+
+                for (let i = 1; i < output.length; i++)
+                    status.modifications.push(...(output[i].trim().split("\n")
+                            .reduce(manageditems_callbacks[i - 1][1], [])));
```
What does the `...` spread do here, I only know this with `[]` and `{}`?

marusak replied:
result of `reduce` is [], and I want to append these to `status.modifications` as separate items and not as one array.

martinpitt commented on pkg/selinux/selinux-client.js:
```diff
@@ -96,6 +102,66 @@ export function init(statusChangedCallback) {
+            }
+        )
+                .catch(e => {
+                    if (e.message === "ValueError: SELinux policy is not managed or store cannot be accessed.") {
```
This is an awfully long and specific string to rely on. Would the `ValueError` be enough?

marusak replied:
I am not sure about this - I don't know what other possible errors can occur. But surely we can go with `ValueError` and then just see, if it may be misleading.

martinpitt commented on pkg/selinux/setroubleshoot-view.jsx:
```diff
@@ -452,6 +455,15 @@ export class SETroubleshootPage extends React.Component {
+        modifications = (
+            <Modifications
+                title={ _("System Modifications") }
+                emptyCaption={ this.props.selinuxStatus.permitted ? _("No System Modifications") : cockpit.format(_("The user <b>$0</b> is not permitted to view system modifications"), permission.user ? permission.user.name : '') }
```
It seems like the permissions rendering and strings could move into the `Modifications` library. You should just pass a boolean whether it's permitted or not.

marusak replied:
The thinking why it is here was that the `Modifications` library should be neutral from any specific strings, so it could be later reused in other places (like firewall or so)

martinpitt commented on test/verify/check-setroubleshoot:
```diff
@@ -143,6 +149,18 @@ class TestSelinux(MachineCase):
         b.click(row_selector + " button.pficon-delete")
         b.wait_not_present(row_selector)
 
+        if "distropkg" in m.image: # Introduced in #11789
```
Please enumerate rhel-8-0-distropkg explicitly here. We'll have e. g. 8-{1,2}-distropkg at some point and don't want to skip all of them.

martinpitt commented on test/verify/check-setroubleshoot:
```diff
@@ -143,6 +149,18 @@ class TestSelinux(MachineCase):
+        b.wait_in_text("#basic-tabs-pf-pane-1 > pre:nth-child(1)", "boolean -D")
+        b.wait_in_text("#basic-tabs-pf-pane-1 > pre:nth-child(1)", "fcontext -D")
+        b.wait_in_text("#basic-tabs-pf-pane-1 > pre:nth-child(1)", "boolean -m -1 zebra_write_config")
+        b.wait_in_text("#basic-tabs-pf-pane-1 > pre:nth-child(1)", "fcontext -a -f a -t samba_share_t -r 's0' '/var/tmp/'")
```
Please also check the "no modifications" vs. "not permitted" cases.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,159 @@
+        super(props);
+
+        this.state = {
+            active_tab: "shell"
```
Where is this being used? Does the `<NavItem>` component know about this property? It was `key` before, and it still says `eventKey`, so shouldn't there be some `key={this.state.active_tab}` somewhere?

marusak replied:
This is not used anywhere at the moment. I just keep it for the future, when we add for example download button we will need to know which tab is selected.

martinpitt replied:
Ack, thank you!

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,159 @@
+        return (
+            <Modal show={this.props.show} className="automation-script-modal">
+                <Modal.Header>
+                    <Modal.Title>{ _("Automation script") }</Modal.Title>
```
Title Case here

martinpitt commented on pkg/selinux/selinux-client.js:
```diff
@@ -96,6 +102,66 @@ export function init(statusChangedCallback) {
+        return result;
+    }
+
+    var getModifications = function() {
```
Still the same weird function syntax.

martinpitt commented on pkg/selinux/selinux-client.js:
```diff
@@ -96,6 +102,66 @@ export function init(statusChangedCallback) {
+        // Use `echo '~~~~~'` as separator, so we don't need to execute multiple commands
+        let script = "semanage export";
+        manageditems.forEach(item => { script += " && echo '~~~~~' && semanage " + item + " --list -C --noheading" });
+        cockpit.script(script, [], { err: 'message', environ: [ "LC_MESSAGE=C" ], superuser: "try" })
```
It's `LC_MESSAGES`

marusak commented:
I've dropped a bit of css as well and used colors from variables.

martinpitt commented:
Interactively testing:

 - For me, reading the modifications took about 5 s (I'm on a busy system, doing some compilation in the background). Until then, it said "No modifications", but it would better show a spinner while the status is still uninitialized.

- The dialog feels a bit bare. ISTM that it could do with some message like "Use this shell script to recreate the current modifications on another system". @andreasn, WDYT?

martinpitt commented:
Also, it's irritatingly hard to actually use the shell script -- Ctrl+A doesn't work at least in Firefox (it selects the entire page), so the only option is to manually mouse-mark the whole script through scrolling. Could we make Ctrl+A work somehow, or show a "Copy to Clipboard" icon?

andreasn commented:
If I shift+reload the page, it takes about 3 seconds before the local modifications show up, and it will say _No System Modifications_, a statement that is actually false (it just haven't figured out that is the case yet). Would it be possible to throw in a spinner while it's figuring that out?

andreasn commented:
> Also, it's irritatingly hard to actually use the shell script -- Ctrl+A doesn't work at least in Firefox (it selects the entire page), so the only option is to manually mouse-mark the whole script through scrolling. Could we make Ctrl+A work somehow, or show a "Copy to Clipboard" icon?

Yeah, I think we should. Do we want to block this PR on that, or do it as a followup?

martinpitt commented:
Honestly I think we should block on a proper copy&paste or "use this shell script" story, otherwise it feels incomplete.

marusak commented:
> Yeah, I think we should. Do we want to block this PR on that, or do it as a followup?

I have already implemented `copy to clipboard` button. Now just thinking where to put it :)

andreasn commented:
> Now just thinking where to put it :)

Hierarchy-wise it would make most sense inside the text field, but that risks covering up content.
Maybe the bottom left corner?

andreasn commented:
With the PF4-styling landing recently on master, the label for _View automation script_ feels very small.
![Screenshot from 2019-07-02 16-59-33](https://user-images.githubusercontent.com/209895/60523310-cac99680-9cea-11e9-9f6b-03f432629ec6.png)

I think this css rule needs to be cut
_.modifications-caption a.modifications-export {
  font-size: 12px;
}_

marusak commented:
> Maybe the bottom left corner?

Like this? I don't like it at all...
![buttonleftdown](https://user-images.githubusercontent.com/12330670/60522970-34957080-9cea-11e9-8134-fb9180e5d662.png)

On the top right I like it more:
![righttopcorner](https://user-images.githubusercontent.com/12330670/60523264-b4bbd600-9cea-11e9-8c42-e252a31b00f2.png)

Or you meant outside?
![otuside](https://user-images.githubusercontent.com/12330670/60523545-36abff00-9ceb-11e9-9d55-cf2267fd3661.png)

(I just very quickly moved the icon around the page, so not much aligning there, just for a broad idea)

andreasn commented:
> Or you meant outside?

Yes, outside works best, sorry for being vague.
Do you have any text with that? The icon by itself feels a bit vague, even though I've seen similar pattern used on GitHub and Nextcloud.

marusak commented:
Getting there? 🤞
![newnewnen](https://user-images.githubusercontent.com/12330670/60527714-11bb8a00-9cf3-11e9-8aa3-8fed7adcbdb0.png)

andreasn commented:
> Getting there? crossed_fingers

That looks great!

marusak commented:
I've added spinner while loading modification.
![loadinguntilendoftime](https://user-images.githubusercontent.com/12330670/60569072-ba5aff80-9d6e-11e9-95fd-ee96448de4c3.png)

I have added also copy to clipboard button: (separate commit for simpler review, can be squashed on merging)
![systemods12](https://user-images.githubusercontent.com/12330670/60569101-c941b200-9d6e-11e9-81ca-f2165a41451d.png)

And made the `View automation script` label bigger:
![systemods11](https://user-images.githubusercontent.com/12330670/60569153-ef675200-9d6e-11e9-9885-064d6273bf65.png)

marusak commented:
Failed tests are unrelated. @martinpitt @andreasn any other comments?

andreasn commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,187 @@
+                <tr className="modification-row">
+                    <td>
+                        <div className="spinner spinner-sm" />
+                        <span>{_("Loading system modification...")}</span>
```
_Modification_ is singular here, even though it's plural everywhere else.
Should probably be _modifications_

andreasn commented:
Looks great in general, and the copy button is great.
I feel there is some feedback missing when it successfully copied though.
Here on github, under _files changed_ when you copy to the clipboard, it changes the little icon to a checkmark for 2-3 seconds after you pressed the copy button. Maybe we should try that as well.
Should we change the text temporary as well? Can we do that without changing the button size?

andreasn commented:
Top one is the current state, the 3 others are ideas for how we can message the copy.

![copied-to-clipboard](https://user-images.githubusercontent.com/209895/60589370-a4166900-9d99-11e9-90d6-b31d08d6e411.png)

andreasn commented:
I kind of like the minimalism of the bottom one.

andreasn commented:
And probably easiest to implement, because it's just a matter of changing the icon on onclick

marusak commented:
I've implemented the last one. Would you please test it and let me know how it feels? It stays for 3s and then switches back (It is very simple, so if you click more times then some older timer could reset the icon sooner - but I think it is fine for this usecase)
![Screenshot from 2019-07-03 21-34-05](https://user-images.githubusercontent.com/12330670/60620152-bebd0200-9dda-11e9-8854-1d274a95beea.png)

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -37,16 +37,33 @@ class ModificationsExportDialog extends React.Component {
         this.setState({ active_tab });
     }
 
+    getScript() {
```
This name is a bit confusing , copyToClipboard() would be a bit clearer. But not a biggie if that's the only outstanding thing.

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -37,16 +37,33 @@ class ModificationsExportDialog extends React.Component {
+                    })
+                    .catch(e => console.error('Text could not be copied: ', e ? e.toString() : ""));
+        } catch (error) {
+            console.error('Text could not be copied: ', error.toString());
```
Why does this need to be handled twice?

Do we expect this to fail in some scenarios, like particular browsers? If so, do we need to change the button then (perhaps with a red X or so) to indicate the failure? Or is that a Should Not Happen™ case?

marusak replied:
One catches when this API is not supported by browser and second one if there are no permissions.
I think this is rather edge case (copying to clipboard works, reading clipboard is usually the problematic one) but surely it could happen. Should Not Happen™ sounds about right, but I can handle it in the UI. @andreasn what do you think and if you think this should be visually shown, what would be the right UI?

martinpitt replied:
Thanks. Indeed if writing clipboard doesn't ask, then it's fine IMHO.

andreasn replied:
I think it's fine to only show it in the console if it's so rare.

martinpitt commented on test/verify/check-setroubleshoot:
```diff
@@ -152,14 +152,32 @@ class TestSelinux(MachineCase):
         if m.image in ["rhel-8-0-distropkg"]: # Introduced in #11789
             return
 
+        b.grant_permissions("clipboardRead", "clipboardWrite")
```
Is that only for the headless browser, or does interactive chromium ask for permission as well? If it does that, then IMHO the "failed to copy" can happen in practice and we should handle this somehow in the UI.

marusak replied:
I could not block copying anyhow - setting clipboard in chrome/ium to Ask/Block and still copying worked. This is mostly there because I paste it into terminal. I'll test it how it works for the copying part for tests.

marusak replied:
But tests need it, otherwise pasting nor copying works. If I understood it correctly, copying to clipboard does not need permissions when triggered by user action (like in `onClick()`) but reading the clipboard needs these permissions even in such case. (From [here](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Interact_with_the_clipboard)). Here in selinux we implement only copying to clipboard.

martinpitt replied:
Thanks. Perhaps this grant_permissions() call could move down in between copying in the changes dialog and pasting into the terminal?

marusak replied:
It could be moved closer where it is really needed, but as I wrote, in tests it is needed also for copying the changes - seems that the browser does not recognize our tests clicking on button as a real user action.

martinpitt replied:
Ack, thanks.

martinpitt commented on test/verify/check-setroubleshoot:
```diff
@@ -152,14 +152,32 @@ class TestSelinux(MachineCase):
+        b.click('.contextMenu .contextMenuOption:nth-child(2)')
+        b.wait_in_text(".xterm-accessibility-tree", "boolean -D")
+        b.wait_in_text(".xterm-accessibility-tree", "boolean -m -1 zebra_write_config")
+        b.wait_in_text(".xterm-accessibility-tree", "fcontext -a -f a -t samba_share_t -r 's0' '/var/tmp/'")
```
Haha, this is a clever way of testing the clipboard :+1:

martinpitt commented on pkg/lib/cockpit-components-modifications.jsx:
```diff
@@ -0,0 +1,173 @@
+                        <tbody>
+                            {this.props.entries.map(entry => <tr className="modification-row" key={entry.split(' ').join('')}><td>{entry}</td></tr>)}
+                        </tbody>
+                    }
```
This could be written a tad more elegantly with a single `{ emptyRow || <tbody>...}`, or maybe calling the variable `row` and assigning one or the other to it? Then this would also make this a bit more readable.
